### DISCUSSION TITLE
OOM-Killer when building tiles

### DISCUSSION THREAD

**Discussion by ahmadshiddiqn** [valhalla/valhalla#3138] [Q&A]
Hello, I'd like to know if there any workaround for this.

So I'm building my country's tiles on the background. When I checked later, the process stop before cleanup stage. I try logging it to a file, and it stop (everytime i retry) on this particular steps:
![image](https://user-images.githubusercontent.com/45712248/121490828-97780200-c9ff-11eb-9719-48ae0453e4f8.png)

I tried looking for what happen, and somehow it got killed by OOM-Killer.
![image](https://user-images.githubusercontent.com/45712248/121491276-ff2e4d00-c9ff-11eb-931f-84dda541daac.png)

I'm not that used in linux environment, and this is the first time I encountered this. This is the environment I used to build the tiles: *captured from `lshw` *I use this on server with Ubuntu 20.04.2 LTS
```
*-memory
          description: System memory
          physical id: 0
          size: 4GiB
     *-cpu
          product: Intel(R) Xeon(R) CPU           E5520  @ 2.27GHz
          vendor: Intel Corp.
          physical id: 1
          bus info: cpu@0
          size: 1717MHz
          capacity: 2268MHz
          width: 64 bits
```
This is what `/usr/bin/time` output:
![image](https://user-images.githubusercontent.com/45712248/121502735-a57f5000-ca0a-11eb-85ec-bdf71183fa43.png)

and this is htop the time before it got killed
![image](https://user-images.githubusercontent.com/45712248/121503391-39e9b280-ca0b-11eb-9c8b-7bd3346b0b06.png)

Any ideas? Thank you!

**Comment by nilsnolde** [comment]
not entirely sure how you get around the RAM limitation for bigger extracts such as indonesia (maybe enable swapping? would be hellish slow though). 4 GB are fine for running valhalla, you should maybe consider adding a builder node which you can use only for building tiles. or do it locally on your laptop (if it has more RAM) and upload the tarred extract to your server. if you go that route, best use docker, either this repo's image or ours: https://github.com/gis-ops/docker-valhalla

**Reply by ahmadshiddiqn** [reply]
Building it on my laptop then upload it is the most understandable way for me lol, but I'll try to look into other suggestions too. Sounds interesting, thank you very much

**Comment by kevinkreiser** [comment]
Yeah exactly. @nilsnolde is right 4GB is pretty small for building tiles, there are a couple data structures which will take up half a gig or more in different places of the tile building process. I have to review the enhancer to see what that could be though. The first thing that comes to mind is that each thread has its own graph reader and each of those has its own in memory tile cache. By default it uses your systems full parallelism which is 8 threads. If you have 4gb shared amongst 8 threads you will need to reduce the tile cache size so they clear memory more often: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L14 by default its a gig and of course 8gb (1gb per thread) is more than your max of 4gb. The other option is to reduce your concurrency but that will probably be significantly slower so i wouldnt try that first. To try that you can change this config option: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L22

**Reply by ahmadshiddiqn** [reply]
I've been thinking about limiting the memory usage but I don't know how. Thank you for the reply, I might try one of the suggestion later
